src/Storages/MergeTree/MergeTreeReadPoolParallelReplicasInOrder.cpp

```diff
@@ -65,21 +71,77 @@ MergeTreeReadTaskPtr MergeTreeReadPoolParallelReplicasInOrder::getTask(size_t ta
     std::lock_guard lock(mutex);
 
     if (task_idx >= per_part_infos.size())
-        throw Exception(ErrorCodes::LOGICAL_ERROR,
-            "Requested task with idx {}, but there are only {} parts",
-            task_idx, per_part_infos.size());
+        throw Exception(
+            ErrorCodes::LOGICAL_ERROR, "Requested task with idx {}, but there are only {} parts", task_idx, per_part_infos.size());
 
     const auto & part_info = per_part_infos[task_idx]->data_part->info;
-    auto get_from_buffer = [&]() -> std::optional<MarkRanges>
+    const auto & data_settings = per_part_infos[task_idx]->data_part->storage.getSettings();
+    auto get_from_buffer = [&,
```

Comment:
let's create a test for mem usage that reproduces the problem

Reply:
I will add some tests